# Logistics Operations — Data Exploration
**Project:** Logistics Capstone  
**Notebook:** 01 — Load & Explore All Sheets  
**Tools:** pandas  

---
**Goal:** Load all 14 sheets, understand their shape, spot any quality issues, and map the relationships before we start merging or analysing anything.

## 1. Import libraries

In [2]:
import pandas as pd

# Show all columns when printing a DataFrame
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 50)
pd.set_option('display.float_format', '{:,.2f}'.format)

print('pandas version:', pd.__version__)

pandas version: 2.3.3


## 2. Load all 14 CSVs

Each table is a separate CSV file in the `data/` folder.
We load them directly into named DataFrames using `pd.read_csv()`.

In [3]:
import pandas as pd
import os

DATA_PATH = '../data/'  # relative path from notebooks/ folder

# Load all 14 CSVs
drivers          = pd.read_csv(DATA_PATH + 'drivers.csv')
trucks           = pd.read_csv(DATA_PATH + 'trucks.csv')
trailers         = pd.read_csv(DATA_PATH + 'trailers.csv')
customers        = pd.read_csv(DATA_PATH + 'customers.csv')
facilities       = pd.read_csv(DATA_PATH + 'facilities.csv')
routes           = pd.read_csv(DATA_PATH + 'routes.csv')
loads            = pd.read_csv(DATA_PATH + 'loads.csv')
trips            = pd.read_csv(DATA_PATH + 'trips.csv')
fuel_purchases   = pd.read_csv(DATA_PATH + 'fuel_purchases.csv')
maintenance      = pd.read_csv(DATA_PATH + 'maintenance_records.csv')
delivery_events  = pd.read_csv(DATA_PATH + 'delivery_events.csv')
safety_incidents = pd.read_csv(DATA_PATH + 'safety_incidents.csv')
driver_metrics   = pd.read_csv(DATA_PATH + 'driver_monthly_metrics.csv')
truck_metrics    = pd.read_csv(DATA_PATH + 'truck_utilization_metrics.csv')

print('All 14 tables loaded.')

All 14 tables loaded.


## 3. Quick overview of every sheet

This loop gives us a one-line summary for each table:  
how many rows, how many columns, and how many null values exist.

In [4]:
# Build a dict on the fly just for the overview loop
all_tables = {
    'DRIVERS':                  drivers,
    'TRUCKS':                   trucks,
    'TRAILERS':                 trailers,
    'CUSTOMERS':                customers,
    'FACILITIES':               facilities,
    'ROUTES':                   routes,
    'LOADS':                    loads,
    'TRIPS':                    trips,
    'FUEL_PURCHASES':           fuel_purchases,
    'MAINTENANCE_RECORDS':      maintenance,
    'DELIVERY_EVENTS':          delivery_events,
    'SAFETY_INCIDENTS':         safety_incidents,
    'DRIVER_MONTHLY_METRICS':   driver_metrics,
    'TRUCK_UTILIZATION_METRICS':truck_metrics,
}

print(f'{"Table":<35} {"Rows":>8} {"Cols":>6} {"Nulls":>8}')
print('-' * 62)
for name, df in all_tables.items():
    nulls = df.isnull().sum().sum()
    print(f'{name:<35} {df.shape[0]:>8,} {df.shape[1]:>6} {nulls:>8,}')

Table                                   Rows   Cols    Nulls
--------------------------------------------------------------
DRIVERS                                  150     12      124
TRUCKS                                   120     11        0
TRAILERS                                 180      9        0
CUSTOMERS                                200      8        0
FACILITIES                                50      9        0
ROUTES                                    58      9        0
LOADS                                 85,410     12        0
TRIPS                                 85,410     12    5,066
FUEL_PURCHASES                       196,442     11    7,868
MAINTENANCE_RECORDS                    2,920     12        0
DELIVERY_EVENTS                      170,820     11        0
SAFETY_INCIDENTS                         170     15        2
DRIVER_MONTHLY_METRICS                 4,464      9        0
TRUCK_UTILIZATION_METRICS              3,312     10        0


## 4. Deep dive — explore each table

For every table we look at:
- `.head()` — first 5 rows to see what data looks like
- `.info()` — column names, data types, non-null counts
- `.describe()` — stats for numeric columns

Run each cell and **add notes** about anything that looks odd.

### 4a. DRIVERS

In [ ]:
print('Shape:', drivers.shape)
drivers.head()

In [ ]:
drivers.info()

In [ ]:
drivers.describe()

### 4b. TRUCKS

In [ ]:
print('Shape:', trucks.shape)
trucks.head()

In [ ]:
trucks.info()

### 4c. TRAILERS

In [ ]:
print('Shape:', trailers.shape)
trailers.head()

### 4d. CUSTOMERS

In [ ]:
print('Shape:', customers.shape)
customers.head()

In [ ]:
customers.describe()

### 4e. FACILITIES

In [ ]:
print('Shape:', facilities.shape)
facilities.head()

### 4f. ROUTES

In [ ]:
print('Shape:', routes.shape)
routes.head()

In [ ]:
routes.describe()

### 4g. LOADS  *(has foreign keys: customer_id, route_id)*

In [ ]:
print('Shape:', loads.shape)
loads.head()

In [ ]:
loads.info()

In [ ]:
loads.describe()

### 4h. TRIPS  *(central table — joins to almost everything)*

In [ ]:
print('Shape:', trips.shape)
trips.head()

In [ ]:
trips.info()

In [ ]:
trips.describe()

### 4i. FUEL_PURCHASES

In [ ]:
print('Shape:', fuel_purchases.shape)
fuel_purchases.head()

In [ ]:
fuel_purchases.describe()

### 4j. MAINTENANCE_RECORDS

In [ ]:
print('Shape:', maintenance.shape)
maintenance.head()

In [ ]:
maintenance.describe()

### 4k. DELIVERY_EVENTS

In [ ]:
print('Shape:', delivery_events.shape)
delivery_events.head()

In [ ]:
delivery_events.info()

### 4l. SAFETY_INCIDENTS

In [ ]:
print('Shape:', safety_incidents.shape)
safety_incidents.head()

In [ ]:
safety_incidents.describe()

### 4m. DRIVER_MONTHLY_METRICS  *(pre-aggregated — use for trends)*

In [ ]:
print('Shape:', driver_metrics.shape)
driver_metrics.head()

In [ ]:
driver_metrics.describe()

### 4n. TRUCK_UTILIZATION_METRICS  *(pre-aggregated — use for fleet trends)*

In [ ]:
print('Shape:', truck_metrics.shape)
truck_metrics.head()

In [ ]:
truck_metrics.describe()

## 5. Null value breakdown — per table

We already saw the *total* nulls per table above.  
This cell shows which *columns* have nulls so we know what to fix in the cleaning notebook.

In [5]:
for name, df in all_tables.items():
    null_cols = df.isnull().sum()
    null_cols = null_cols[null_cols > 0]
    if len(null_cols) > 0:
        print(f'\n--- {name} ---')
        print(null_cols.to_string())

print('\nDone. Any table not listed above has zero nulls.')


--- DRIVERS ---
termination_date    124

--- TRIPS ---
driver_id     1714
truck_id      1672
trailer_id    1680

--- FUEL_PURCHASES ---
truck_id     3880
driver_id    3988

--- SAFETY_INCIDENTS ---
truck_id     1
driver_id    1

Done. Any table not listed above has zero nulls.


## 6. Verify key relationships

Before merging, we check that foreign keys actually match between tables.  
If the overlap is 100%, the join will be clean. If not, we have orphan records to investigate.

Think of this as the pandas version of a SQL referential integrity check.

In [6]:
def check_fk(child_df, child_col, parent_df, parent_col, label):
    """Check how many FK values in child_df exist in parent_df."""
    child_vals  = set(child_df[child_col].dropna())
    parent_vals = set(parent_df[parent_col].dropna())
    matched     = child_vals & parent_vals
    pct         = len(matched) / len(child_vals) * 100 if child_vals else 0
    status      = 'OK' if pct == 100 else 'WARN'
    print(f'[{status}] {label}: {pct:.1f}% of {len(child_vals):,} values matched')

check_fk(loads,           'customer_id', customers,  'customer_id', 'loads -> customers')
check_fk(loads,           'route_id',    routes,     'route_id',    'loads -> routes')
check_fk(trips,           'load_id',     loads,      'load_id',     'trips -> loads')
check_fk(trips,           'driver_id',   drivers,    'driver_id',   'trips -> drivers')
check_fk(trips,           'truck_id',    trucks,     'truck_id',    'trips -> trucks')
check_fk(trips,           'trailer_id',  trailers,   'trailer_id',  'trips -> trailers')
check_fk(fuel_purchases,  'trip_id',     trips,      'trip_id',     'fuel_purchases -> trips')
check_fk(maintenance,     'truck_id',    trucks,     'truck_id',    'maintenance -> trucks')
check_fk(delivery_events, 'trip_id',     trips,      'trip_id',     'delivery_events -> trips')
check_fk(delivery_events, 'facility_id', facilities, 'facility_id', 'delivery_events -> facilities')
check_fk(safety_incidents,'trip_id',     trips,      'trip_id',     'safety_incidents -> trips')

[OK] loads -> customers: 100.0% of 200 values matched
[OK] loads -> routes: 100.0% of 58 values matched
[OK] trips -> loads: 100.0% of 85,410 values matched
[OK] trips -> drivers: 100.0% of 124 values matched
[OK] trips -> trucks: 100.0% of 92 values matched
[OK] trips -> trailers: 100.0% of 180 values matched
[OK] fuel_purchases -> trips: 100.0% of 76,939 values matched
[OK] maintenance -> trucks: 100.0% of 120 values matched
[OK] delivery_events -> trips: 100.0% of 85,410 values matched
[OK] delivery_events -> facilities: 100.0% of 50 values matched
[OK] safety_incidents -> trips: 100.0% of 170 values matched


## 7. Notes & observations

### Dataset summary
DRIVERS                —  150 rows,  12 cols,  124 nulls
TRUCKS                 —  120 rows,  11 cols,    0 nulls
TRAILERS               —  180 rows,   9 cols,    0 nulls
CUSTOMERS              —  200 rows,   8 cols,    0 nulls
FACILITIES             —   50 rows,   9 cols,    0 nulls
ROUTES                 —   58 rows,   9 cols,    0 nulls
LOADS                  — 85,410 rows, 12 cols,   0 nulls
TRIPS                  — 85,410 rows, 12 cols, 5,066 nulls
FUEL_PURCHASES         — 196,442 rows, 11 cols, 7,868 nulls
MAINTENANCE_RECORDS    —  2,920 rows, 12 cols,   0 nulls
DELIVERY_EVENTS        — 170,820 rows, 11 cols,  0 nulls
SAFETY_INCIDENTS       —  170 rows,  15 cols,    2 nulls
DRIVER_MONTHLY_METRICS —  4,464 rows,  9 cols,   0 nulls
TRUCK_UTILIZATION_METRICS — 3,312 rows, 10 cols, 0 nulls

### Null columns
- drivers.termination_date     → 124 nulls (drivers still employed — fill with "Active")
- trips.driver_id              → 1,714 nulls (unassigned trips — drop for analysis)
- trips.truck_id               → 1,672 nulls (unassigned trips — drop for analysis)
- trips.trailer_id             → 1,680 nulls (unassigned trips — drop for analysis)
- fuel_purchases.truck_id      → 3,880 nulls (drop for truck-level fuel analysis)
- fuel_purchases.driver_id     → 3,988 nulls (drop for driver-level fuel analysis)
- safety_incidents.truck_id    → 1 null (drop row)
- safety_incidents.driver_id   → 1 null (drop row)

### FK integrity
All 11 relationships checked — 100% matched. No orphan records.

### Issues to fix in 02_cleaning.ipynb
1. Fill termination_date nulls with "Active"
2. Drop trips rows where driver_id / truck_id / trailer_id is null
3. Drop fuel_purchases rows where truck_id or driver_id is null
4. Drop the 2 safety_incidents rows with nulls

---
**Next notebook:** `02_cleaning.ipynb` — fix nulls, data types, and prepare merge-ready tables.